### PR TITLE
Fix remote reuse bugs

### PR DESCRIPTION
**Pull request by SparkSnail**

### CODE REVIEW

File: tools/nni_trial_tool/trial.py [hunk 1]
```diff
@@ -137,10 +137,13 @@ def is_running(self):
...
-                nni_log(LogType.Info, "%s: killing trial" % self.name)
-                for child in psutil.Process(self.process.pid).children(True):
-                    child.kill()
-                self.process.kill()
+                try:
+                    nni_log(LogType.Info, "%s: killing trial" % self.name)
+                    for child in psutil.Process(self.process.pid).children(True):
+                        child.kill()
+                    self.process.kill()
+                except Exception as ex:
```
Comment (liuzhe-lz) [inline tools/nni_trial_tool/trial.py:145]: Can we catch a more specific bug?

File: tools/nni_trial_tool/trial.py [hunk 2]
```diff
@@ -137,10 +137,13 @@ def is_running(self):
...
         if trial_id == self.id or trial_id is None:
             if self.process is not None:
-                nni_log(LogType.Info, "%s: killing trial" % self.name)
-                for child in psutil.Process(self.process.pid).children(True):
-                    child.kill()
-                self.process.kill()
+                try:
+                    nni_log(LogType.Info, "%s: killing trial" % self.name)
+                    for child in psutil.Process(self.process.pid).children(True):
+                        child.kill()
```
Comment (squirrelsc) [inline tools/nni_trial_tool/trial.py:143]: If there is try catch, it's better to try catch for each kill. So that one fail won't effect others.

File: tools/nni_trial_tool/trial.py [hunk 3]
```diff
@@ -137,10 +137,13 @@ def is_running(self):
...
+                    self.process.kill()
+                except Exception as ex:
+                    nni_log(LogType.Error, "kill trial %s failed: %s " % (trial_id, str(ex)))
```
Comment (squirrelsc) [inline tools/nni_trial_tool/trial.py:146]: It likes a clean up, don't need error level log. debug or info is enough.

File: src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts [hunk 4]
```diff
@@ -137,40 +137,37 @@ export class RemoteEnvironmentService extends EnvironmentService {
...
+                            environment.setStatus('FAILED');
                         }
+                        this.releaseEnvironmentResource(environment);
                     }
-                } catch (error) {
-                    this.releaseEnvironmentResource(environment);
-                    this.log.error(`Update job status exception, error is ${error.message}`);
                 }
             }
+        } catch (error) {
```
Comment (QuanluZhang) [inline src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:168]: no need to release environment resource here, why?

**Comment by QuanluZhang** [comment]
@SparkSnail could you briefly explain why your changes fix the problem?

File: src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts (hunk 4 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:168]: There is a case that environment is submitted, but it starts slowly and hasn't start process and create pid file, the system call refresh function to read pid file, it will cause `no such file` exception.

Comment (squirrelsc) [reply src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:168]: I see, it looks you don't check this file. Check if environment.isRunnerReady, then check the file. It will depend on first initialized message. And you set env status to running is too earlier. In remote, it's better to wait isRunnerReady first, then check file status, and set to running, success or failed.

File: tools/nni_trial_tool/trial.py (hunk 2 above)
Comment (SparkSnail) [reply tools/nni_trial_tool/trial.py:143]: This error is used to catch psutil.Process(), not child kill. In some kind of scene, trial has already exited, and kill() command is sent later. Will throw process not exist error.

File: tools/nni_trial_tool/trial.py (hunk 1 above)
Comment (SparkSnail) [reply tools/nni_trial_tool/trial.py:147]: ok, add a NoSuchProcess.

File: tools/nni_trial_tool/trial.py (hunk 3 above)
Comment (SparkSnail) [reply tools/nni_trial_tool/trial.py:148]: psutil.NoSuchProcess is expected exit issue, use info level. For other kinds of unexpected issue, I think use error level is better.

File: src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts (hunk 4 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:168]: no, if the environment is failed to start, the isRunnerReady will always be false, but we need to refresh env status to failed here.

Comment (squirrelsc) [reply src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:168]: How do you know it failed to start? You may can wait the pid file when initializing, instead of set env to running directly.

Comment (SparkSnail) [reply src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts:174]: Check process return code to detect if env is failed to start. Added detecting logic for pid file exist.

File: src/nni_manager/training_service/reusable/trialDispatcher.ts [hunk 5]
```diff
@@ -664,17 +664,16 @@ class TrialDispatcher implements TrainingService {
     }
 
     private releaseEnvironment(trial: TrialDetail): void {
-        if (undefined === trial.environment) {
-            throw new Error(`TrialDispatcher: environment is not assigned to trial ${trial.id}, and cannot be released!`);
```
Comment (squirrelsc) [inline src/nni_manager/training_service/reusable/trialDispatcher.ts:668]: What's reason to remove the check? it helps to find unexpected behavior

Comment (SparkSnail) [reply src/nni_manager/training_service/reusable/trialDispatcher.ts:668]: IT find a case that assessor report two kill command continuous for a trial, a trial released environment, then release again will throw exception.

**Review by QuanluZhang — APPROVED** [review]

**Review by chicm-ms — APPROVED** [review]

File: src/nni_manager/training_service/remote_machine/osCommands.ts [hunk 6]
```diff
@@ -29,6 +29,7 @@ abstract class OsCommands {
     public abstract extractFile(tarFileName: string, targetFolder: string): string;
     public abstract executeScript(script: string, isFile: boolean): string;
     public abstract addPreCommand(preCommand: string | undefined, command: string | undefined): string | undefined;
+    public abstract fileExistCommand(filePath: string): string | undefined;
```
Comment (squirrelsc) [inline src/nni_manager/training_service/remote_machine/osCommands.ts:32]: fileExists is enough. Command is for command related, but this one is not related.

**Review by squirrelsc — APPROVED** [review]
